### PR TITLE
Added more options for homotopy-based initialization of limiters

### LINKED ISSUE #2270: IdealizedOpAmpLimted has issues
The new part Modelica.Electrical.Analog.Ideal.IdealizedOpAmpLimted has a few issues.

A minor issue is that the name has a typo; **Limted** should be **Limited**. 

A serious bug is that the external supply pins, s_p and s_n, which are enabled when the boolean input useSupply is true, do not work because the equation that controls the output is in terms of Vps and Vns, the parameters, rather than vps and vns, the variables that should have been used.

From a performance viewpoint, the use of smooth(0, min(vps, max(vns, V0*v_in))) to generate the output, while convenient, is not ideal. The smooth(0, .) asserts that the output is continuous.  That is not necessarily the case if v_in is not continuous, which can happen if the input is derived from a step signal or the circuit being modeled employs positive feedback.  The bigger issue is that the use of smooth with min and max precludes the creation of events which would otherwise improve the engine performance at the transition to limiting.  In a typical circuit where the external gain, set by the feedback, is much smaller than the open-loop gain (V0), the change in slope of the controlled signal as the circuit transitions to and from limiting is a few orders of magnitude. Without an event this can be problematic for some integrators to maintain error control.  A better solution is to use a condition:

`v_out = if     AtNegRail then vns
                     elseif AtPosRail then vps
                     else                   V0*v_in;`

where AtNegRail and AtPosRail are boolean variables defined as

`Boolean AtNegRail = V0*v_in < vns "True when output is at the negative rail";`
`Boolean AtPosRail = V0*v_in > vps "True when output is at the positive rail";`

The inequalities can be put directly into the conditions of the if-expression, however, declaring them as probeable variables provides a convenient means to detect that the opamp is at the rails, a normally undesirable condition.

### PR DESCRIPTION
**Pull request by casella**
With MSL 3.2.2, the behaviour of limiters in the Modelica.Blocks library was slightly changed, so that if homotopy-based initialization was used, a linear behaviour y = u was always assumed.

Although this obviously makes the simplified model more linear, and thus potentially easier to solve, it can be problematic if the model receiving the output of the limiter breaks for values outside the limiting range and one still wants to use homotopy.

Also, in some cases it is known a priori that in the initial conditions the output of the limiter is stuck at the upper or lower limit. In this case, it is much better to use y = uMin or y = uMax as output in the simplified model.

This pull requests solves all  these issues by allowing to choose among several options that span all these cases, using an enumeration parameter in the Advanced tab.

The change is 100% backwards compatible, as the behaviour with the default values of the newly added parameters is exactly the same as with MSL 3.2.2

close #2270

### CODE REVIEW

File: Modelica/Blocks/Nonlinear.mo [hunk 1]
```diff
@@ -31,6 +49,20 @@ as long as the input is within the specified upper and lower
...
 as output.
 </p>
+<p>
+The parameter <code>homotopyType</code> in the Advanced tab specifies the
+simplified behaviour if homotopy-based initialization is used:
+<ul>
+<li><code>noHomotopy</code>: the actual expression with limits is used</li>
+<li><code>linear</code>: a linear behaviour y = u is assumed (default option)</li>
+<li><code>uMax</code>: it is assumed that the output is stuck at the upper limit u = uMax</li>
+<li><code>uMin</code>: it is assumed that the output is stuck at the lower limit u = uMax</li>
```
[the remark this answers is not in the document]
Comment (casella) [reply Modelica/Blocks/Nonlinear.mo]: Of course, that was a stupid copy&paste mistake

File: Modelica/Blocks/Types.mo [hunk 2]
```diff
@@ -121,6 +121,19 @@ initialization definition.
...
+      uMax "Simplified model: y = uMax",
+      uMin "Simplified model: y = uMin")
+    "Enumeration defining use of homotopy in limiter components" annotation (Evaluate=true);
+
+  type VariableLimiterHomotopy = enumeration(
+      noHomotopy "Simplified model = actual model",
+      linear "Simplified model: y = u",
+      fixed "Simplified model: y = ySimplified")
+    "Enumeration defining use of homotopy in variable limiter components" annotation (Evaluate=true);
+
```
[the remark this answers is not in the document]
Comment (casella) [reply Modelica/Blocks/Types.mo]: Sure, will do

**Comment by beutlich** [comment]
Rescheduling milestone since MSL v3.2.3 is feature-complete, i.e., no more enhancements are allowed. Furthermore, test models for the other homotopy options are required.

**Comment by beutlich** [comment]
@casella Thanks for exaplanation that this PR rather is considered to be a bug-fix. In that case it is valid to be included in the MSL v3.2.3. Please provide a test model (for ModelicaTest.Blocks).

I still believe, we also should add the homotopy choices to the limiter in Modelica.Electrical.Analog.Examples.OpAmps.SignalGenerator. What do you think?

**Comment by casella** [comment]
@beutlich, I added the requested test cases to ModelicaTest in bc99284. Can you proceed?

**Comment by beutlich** [comment]
@casella What about the limiter in Modelica.Electrical.Analog.Examples.OpAmps.SignalGenerator?

**Comment by casella** [comment]
@beutlich, regarding IdealizedOpAmpLimited and IdealOpAmpLimited,

These models not only include a saturation on the output voltage, but they also change the relationship between the input voltages according to that. Basically, if the output voltage is not saturated, then the two input voltage are bound to be equal, otherwise they are not. As you can see, this involves an auxiliary variable s. The logic is a bit the same as the one of the ideal diode.

I'm not sure how this model could be improved by using homotopy, because I'm not an expert of electrical models (my degree in electronic engineering is now almost 25 years old...). Anyway, it's way beyond the scope of this PR, which is only meant to provide Limiter blocks that can be used with homotopy.

**Comment by casella** [comment]
@beutlich, regarding LimPID, I guess propagating homotopyType is a good idea. I'lll do that and build a couple of test cases then I'll get back to you

**Comment by casella** [comment]
In fact, LimPID is still using the obsolete limitsAtIniti parameter, which is propagated to the limiter but actually not used sinces 3.2.2. I'll move that to a Dummy tab as for Limiter and put homotopyType instead

**Comment by AHaumer** [comment]
I agree wirh Francesco, Ideal(ized)OpAmpLimited need some more investigation.
Shift that issue to the next release.

**Comment by AHaumer** [comment]
Regarding the LimPID (which is a commonly used block) I apperciate Franseco's idea propagating homotopyType.

**Comment by casella** [comment]
Updated implementation of LimPID and added relevant test cases to ModelicaTest in 153352f.
Also updated the documentation of LimPID.

**Comment by beutlich** [comment]
I do not quite understand why there are two enumration types InitPIDHomotopy and LimiterHomotopy with the same semantics of the alternatives. I'd simply reuse LimiterHomotopy in LimPID and generalize the alternative names and descriptions, say NoHomotopy, Linear, UpperLimit, LowerLimit.

**Comment by casella** [comment]
@beutlich I just could not find general enough names for the enumeration, but I guess your idea is good. I'll take care of that later today.

**Comment by casella** [comment]
Done. As with research work, I hate reviews, but they help a lot improving the quality of the end product. That said, I hope this can now be merged.

**Comment by casella** [comment]
@beutlich: if you are fine with this, I will also add this feature to LimIntegrator.

Regarding your comment about "I'd move the Dummy tab right to the Advanced tab, hang on", I'm not sure what you mean. In Dymola, the Dummy tab is already to the right of the Advanced tab, I guess they are just alphabetically ordered.

**Comment by beutlich** [comment]
OK, need to check, why SimulationX has Dummy first
grafik

**Comment by casella** [comment]
I analyzed LimIntegrator, but the story is quite different from Limiter, because this model also has a state variable.

A naive implementation of LimIntegrator where I apply homotopy to der(y) as I did for Limiter ends up giving singular initial equations if I select initType=steadyState and then homotopyType=Fixed, because der(y) = 0 will be stated twice.

What would need to be done is to completely rewrite this model by introducing an anti-windup strategy in the same way of LimPID. But in fact, in order to get this behaviour, one can just use LimPID and only turn on integral action. There's no need to rewrite LimIntegrator to have the same effect.

Bottom line, as far as I understand, rewriting LimIntegrator to use homotopy goes way beyond the scope of this PR.

**Comment by beutlich** [comment]
I moved parameters homotopyType and limitsAtInit to look consistent in Limiter, VariableLimiter and LimPID. Hope you agree.

**Comment by casella** [comment]
@beutlich do we need another review? I lost track after so many changes

**Comment by dietmarw** [comment]
Yes by either @AHaumer or @MartinOtter

**Comment by beutlich** [comment]
@HansOlsson 9d7df7b4fe18e5141bcb6d7809de9a09a8f40485 and 68f1beb044a1ea57627f91eaa2fe6ee7df65904b should fix the Modelica.Electrical.Analog.Examples.OpAmps.Multivibrator issue of https://github.com/modelica/ModelicaStandardLibrary/issues/2270#issuecomment-396184954 while still generating events for Modelica.Electrical.Analog.Examples.OpAmps.SignalGenerator. Can you please cross-check and give your review.

**Comment by beutlich** [comment]
@HansOlsson Can you please check if this fixes all examples in Modelica.Electrical.Analog.Examples.OpAmps?

**Comment by beutlich** [comment]
> I would prefer that strict was as default true - corresponding to the previous version.

You mean the IdealizedOpAmpLimted.strict?

**Comment by HansOlsson** [comment]
> And touch each of the example to have strict modified to false, at least if we want to see events (as in Modelica.Electrical.Analog.Examples.OpAmps.SignalGenerator)?

I would skip that for the examples in MSL - since I the events are not needed. 
However, the external code that triggered the change should have strict=false.

**Comment by beutlich** [comment]
According to #2017, Modelica.Electrical.Analog.Examples.OpAmps.SignalGenerator should generate events, right? Thus strict=false.

**Comment by HansOlsson** [comment]
> According to #2017, Modeling of Modelica.Electrical.Analog.Examples.OpAmps.SignalGenerator, Modelica.Electrical.Analog.Examples.OpAmps.SignalGenerator should generate events, right? Thus strict=false.

According to that issue, yes, but based on the problem we had with examples in MSL I  thought that backwards compatibility was more important, so to get #2017 to run you set strict=false.
